test-data/unit/check-optional.test

```diff
@@ -1040,3 +1040,223 @@ x: Optional[List[int]]
 if 3 in x:
     pass
 
+[case testNarrowedVariableInNestedFunctionBasic]
+from typing import Optional
+
+def can_narrow(x: Optional[str]) -> None:
+    if x is None:
+        x = "a"
+    def nested() -> str:
+        return reveal_type(x)  # N: Revealed type is "builtins.str"
+    nested()
+
+def foo(a): pass
+
+class C:
+    def can_narrow_in_method(self, x: Optional[str]) -> None:
+        if x is None:
+            x = "a"
+        def nested() -> str:
+            return reveal_type(x)  # N: Revealed type is "builtins.str"
+        # Reading the variable is fine
+        y = x
+        with foo(x):
+            foo(x)
+        for a in foo(x):
+            foo(x)
+        nested()
+
+def can_narrow_lambda(x: Optional[str]) -> None:
+    if x is None:
+        x = "a"
+    nested = lambda: x
+    reveal_type(nested())  # N: Revealed type is "builtins.str"
+
+def cannot_narrow_if_reassigned(x: Optional[str]) -> None:
+    if x is None:
+        x = "a"
+    def nested() -> str:
+        return x  # E: Incompatible return value type (got "Optional[str]", expected "str")
+    if int():
+        x = None
+    nested()
+
+x: Optional[str] = "x"
+
+def cannot_narrow_top_level() -> None:
```

Comment:
The actual narrowing is missing here. Until #2008 is fixed this would be optional despite r.h.s. So you need to write 
```python
x: Optional[str]
if x is None:
    x = "x"
```
here as well.